03-Binary-Search-Trees/01-Binary-Search-Trees.md

```diff
@@ -875,15 +875,16 @@ As an optional extra challenge, you can attempt to implement `delete`  yourself
 
 ```py
 def delete(self, key):
-    # if the tree is empty, exit the function
-    if not self.root:
-        return
-
     # call our recursive helper on the root
     self.root = self.delete_helper(self.root, key)
 
 # Recursive helper function
 def delete_helper(self, current_root, key):
+    # if the current node is None, the tree is empty 
+    # or the key could not be found
+    if not current_root:
+        return None
```

Comment:
This does feel like it helps the consistency of the operation to me.

Do we need to make a distinction of the tree being empty? An empty tree still means the key cannot be found.

I'm glad this small change also fixes our (presumably unintentional) crash, but should we maybe take an approach more like remove from a liist, which raises a ValueError if we try to remove something that doesn't exist? Though looking at our version of deleting (at a location) from a linked list, we also silently ignore deleting a node that doesn't exist, so this delete is consistent with that approach, and I'm fine with this as is.

But it might be something to consider for our implementations whether we should communicate than an attempted operation could not succeed.

Reply:
Agreed this is worth a larger discussion, thanks for filing the ticket to think about what we want to include and how we're structuring our node examples.